Adding `codec=` option to `st.characters()`

Cheukting opened the pull request:
Attempt to add two options of `codec=`: `ascii` and `utf-8` to `st.characters()` to limit the character generated.

See #1664

Cheukting commented:
@Zac-HD I am a bit confused, I assume Python work in utf-8 by default? Reading this https://github.com/HypothesisWorks/hypothesis/blob/1b2375ba2765a23b610c5bbf03dca5539066d3a5/hypothesis-python/src/hypothesis/strategies/_internal/strings.py#L70 it seems anything out of the code range of utf-8 would not be generated by default?

Maybe I am misunderstanding something, please let me know.

honno commented on hypothesis-python/src/hypothesis/strategies/_internal/core.py:
```diff
@@ -529,6 +529,7 @@ def characters(
     min_codepoint: Optional[int] = None,
     max_codepoint: Optional[int] = None,
     whitelist_characters: Optional[Sequence[str]] = None,
+    codec: Optional[str] = None,
```
Would literals be fine here?

```suggestion
    codec: Optional[Literal["ascii", "utf8"]] = None,
```

Zac-HD commented:
Unfortunately [naming things is hard](https://martinfowler.com/bliki/TwoHardThings.html), and the term _character_ is both vague and overloaded.  To lay things out:

- `st.characters()` generates Python strings, each of which consists of a single Unicode _codepoint_.
  - I agree that [codepoints aren't a great abstraction](https://manishearth.github.io/blog/2017/01/14/stop-ascribing-meaning-to-unicode-code-points/), but that's how Python iterates over strings so that's what the `st.characters()` strategy generates.  If I was writing code that handled (extended) grapheme clusters or text rendering I'd want a custom strategy though 😰 
- so a Python `str` is a sequence of codepoints, while a `bytes` object is a sequence of... bytes (aka _octets_).
- a _codec_ is a particular mapping between codepoints and short bytestrings.  However, it happens to be the case that no standard codec can encode every codepoint in strict mode 😱
  - ASCII can only encode codepoints [0 .. 127], because it predates Unicode.
  - `utf-8` can encode almost everything.  However, [_surrogate characters_](https://unicodebook.readthedocs.io/unicode_encodings.html#surrogates) (an unfortunate legacy from `utf-16`) are [not allowed in `utf-8`-encoded strings, nor other codecs](https://unicodebook.readthedocs.io/issues.html#non-strict-utf-8-decoder-overlong-byte-sequences-and-surrogates), by default.
    - 'non-strict' encoders often write them anyway, as with [`"...".encode(errors="surrogatepass")`](https://stackoverflow.com/questions/38147259/how-can-i-convert-surrogate-pairs-to-normal-string-in-python) in [Python](https://docs.python.org/3/library/codecs.html#error-handlers).
    - This motivated the default `alphabet=st.characters(blacklist_categories=['Cs'])` default in `st.text()`, so that generated strings could be encoded by Python's default encoding; while `st.characters()` would be complete for the true Unicode pedants.

Which means:

- `characters(codec="ascii")` will be a more readable alias for `characters(min_codepoint=0, max_codepoint=127))` (plus handling other arguments which might be passed, of course)
- `characters(codec="utf-8")` will be a more readable alias for `characters(blacklist_categories=['Cs']))`
- the docstring is going to be tricky to get right, though I'm happy to help of course

Zac-HD commented on hypothesis-python/src/hypothesis/strategies/_internal/core.py:
```diff
@@ -566,6 +576,18 @@ def characters(
+    if codec == "ascii":
+        if (max_codepoint is None) or (max_codepoint > 127):
+            max_codepoint = 127
+    elif codec == "utf-8":
```
> Non-overlapping equality check (left operand type: "Literal['utf8'] | None", right operand type: "Literal['utf-8']")  [comparison-overlap]

I think this is just a mypy bug and a `# type: ignore` would be appropriate.

Zac-HD commented on hypothesis-python/tests/nocover/test_characters.py:
```diff
@@ -32,3 +32,13 @@ def test_arbitrary_blacklist(data):
     assert c not in blacklist
+
+
+@given(st.characters(codec="ascii"))
+def test_ascii_codec(c):
+    assert c.encode(encoding="ascii").decode(encoding="ascii") == c
+
+
+@given(st.characters(codec="utf-8"))
+def test_utf8_codec(c):
+    assert c.encode().decode() == c
```
We'll want to move these to `/cover/` tests so that we cover the new logic!  I'd also write them with a pytest.mark.parametrize over the different codec names, and then use the `st.data()` strategy to draw from strategies inside the test.

Zac-HD commented on hypothesis-python/src/hypothesis/strategies/_internal/core.py:
```diff
@@ -549,6 +551,14 @@ def characters(
       that list will be not be produced. Any overlap between
       ``whitelist_characters`` and ``blacklist_characters`` will raise an
       exception.
+    - ``codec`` can accept `ascii`, `utf-8` or None as an option. While None
+      will not have any extra limitations, `ascii` limit characters to be valid
+      ascii (i.e. codepoints 0 to 127) and `utf-8` will include most characters
+      except the surrogate characters. In another word,
+      `characters(codec="ascii")`` will be equivalent to
+      `characters(min_codepoint=0, max_codepoint=127))` and
+      `characters(codec="utf-8")`` will be equivalent to
+      `characters(blacklist_categories=['Cs']))`.
```
I'd omit the implementation details, and try writing it as if codecs _in general_ are supported and link to [these docs](https://docs.python.org/3/library/codecs.html#encodings-and-unicode); then note that currently only ascii and utf-8 are implemented.

Zac-HD reviewed (commented):
Shaping up nicely!  

Along with the comments below, I think we'll want to think carefully about how the `codec=` argument interacts with the others - and document what we decide.  I'm inclined to say that it should be the _last_ constraint applied, excluding all characters that can't be encoded using that codec; and it should be an error to pass `whitelist_characters` which are not so encodable.

Zac-HD reviewed (approved):
This ended up being so tricky to review that I actually couldn't do that without trying to write code, and at that point it seemed easier to just push some commits rather than leaving incredibly-detailed nitpicky comments 😅 

At a high level, this was already pretty much done - I moved around the existing tests a bit and changed the interaction with `whitelist_categories` and `whitelist_characters` based on some local experiments, and then noticed that the existing type annotations were too strict and fixed those too.  I'll get the tests passing soon, hopefully...

Thanks again for your contributions Cheuk, and sorry it took me so long to review ❤️

Cheukting commented:
Thank you for bringing this to the finish line @Zac-HD you are amazing as
always 😊

On Tue, 29 Aug 2023 at 19:39, Zac Hatfield-Dodds ***@***.***>
wrote:

> ***@***.**** approved this pull request.
>
> This ended up being so tricky to review that I actually couldn't do that
> without trying to write code, and at that point it seemed easier to just
> push some commits rather than leaving incredibly-detailed nitpicky comments
> 😅
>
> At a high level, this was already pretty much done - I moved around the
> existing tests a bit and changed the interaction with whitelist_categories
> and whitelist_characters based on some local experiments, and then
> noticed that the existing type annotations were too strict and fixed those
> too. I'll get the tests passing soon, hopefully...
>
> Thanks again for your contributions Cheuk, and sorry it took me so long to
> review ❤️
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/HypothesisWorks/hypothesis/pull/3711#pullrequestreview-1601056086>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG3N26PPQAYZMPMRK7KHLDLXXYZNLANCNFSM6AAAAAA3G46HBQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
